### DISCUSSION TITLE
Grule does not evaluate rules when sending many requests/second.

### DISCUSSION THREAD

**Discussion by tmsl31** [hyperjumptech/grule-rule-engine#449] [Q&A]
Hello,
**Problem**
I am using grule as a rule engine 'inside' an HTTP API using GIN, calling it when referencing a certain endpoint. When dealing with a few requests per second (let's say 100), the system is working properly. To determine that it is running correctly I created 1,000 rules with known desired output. However, when increasing the number of r/s the engine seems to start failing in terms of rule evaluation and I get the same input fact without the result of a rule evaluation.

**Approach**
Reading the discussions I found [this post](https://github.com/hyperjumptech/grule-rule-engine/discussions/221) that seemed useful. I decided to implement the third approach given in the accepted answer, creating a pool of engine instances. However, I'm still experiencing the same situation. I'd like to ask for any ideas regarding this problem and how could I avoid 'running over' the engine.

**Some details about code**
- I get a knowledgeBase instance when starting the program execution to avoid loading the rules when evaluating the facts.
- I create a pool of engines. I implement the following methods. When starting the program I call the NewEnginePool function using different values of size (the number of engines)

```
type EnginePool struct {
	pool chan *engine.GruleEngine
}

func NewEnginePool(size int) *EnginePool {
	pool := make(chan *engine.GruleEngine, size)
	for i := 0; i < size; i++ {
		pool <- NewGruleEngine()
	}
	return &EnginePool{pool: pool}
}

func (p *EnginePool) Acquire() *engine.GruleEngine {
	return <-p.pool
}

func (p *EnginePool) Release(engine *engine.GruleEngine) {
	p.pool <- engine
}
```
- I implement an endpoint that calls the processRequest method that should manage the rule execution and the use of the different engines.
```
func processRequest(c *gin.Context) {
	// Read the information.
	var newFact MyFact
	err := c.ShouldBindJSON(&newFact)
	// Acquire an engine instance from the pool
	engine := enginePool.Acquire()
	// Avoid returning the engine instance.
	defer enginePool.Release(engine)
	// Process the request using the acquired engine
	executeEngine(engine, &newFact)
}
```
Thanks in advance for any insight.
Kind regards,

**Comment by tlaraMQ** [comment]
Is there any timeout inside the grule engine or something else that could explain this behaviour?
